packages/flutter/lib/src/services/text_metrics.dart

```diff
@@ -0,0 +1,38 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'text_editing.dart';
+
+/// A read-only interface for accessing visual information about the
+/// implementing text.
+abstract class TextMetrics {
+  /// {@template flutter.services.TextMetrics.getLineAtOffset}
+  /// Return a [TextSelection] containing the line of the given [TextPosition].
+  /// {@endtemplate}
+  TextSelection getLineAtOffset(String text, TextPosition position);
```

Comment:
IMO this is going to be error prone. `TextMetrics`'s layout information is highly dependent on its text, so being able to provide a potentially different String is a bit confusing. 

> enforce what text it's using

Yeah but we can't layout a plain `String`. It has to be styled and you need other inputs for instance the incoming box constraints. Also other methods in the interface don't seem to take a String? All `TextMetrics` implementations are likely going to have states.

And for `RenderEditable.getLineAtOffset`, getting the length of the obscured text should be an implementation detail instead of being exposed as a parameter I think.

Reply:
You're right, this parameter makes the multiple sources of truth thing more confusing, not less.  What we need to do is ensure that the instance of TextMetrics etc. are consistent with the state they're operating on.  I'll remove the parameter and use `_plainText`.